Cli update

lhca opened the pull request:
An updated man page, calcurse(1).

The man-page is mostly concerned with the CLI, and going through it led to some CLI bug fixes as well. There are also two interface changes:

- the day range for queries (`--from`, `--to`) is now inclusive; previously the end day, `--to`, was not included
- the input date format is taken from the configuration file, but may be overridden by a new long option `--input-datefmt`; time (of day) input has been eliminated, but has apparently never been used

The last commit (Safety exit and read-only mode) is put here out of convenience. It has an update and fix for https://github.com/lfos/calcurse/commit/2339d78cbd562319473a34a36390378df64ade00 and is not related to CLI or the other commits.

lfos commented:
I guess we should also update the test suite and enforce the new documented behavior?

lhca commented:
There were only two failed test and they have been updated (https://github.com/lfos/calcurse/pull/162/commits/91930248fd8a0197b297d9530355ec74e583e10e) so that the command line input format agrees with the `conf` file.

And I had to change the test configuration file as well (and set `general.systemdialogs=yes`) even though previous Travis builds succeeded without it. How can that be?

I tried to figure out a way to make the proposed `calcurse(1)` available in HTML form, but couldn't find one (I don't have a home page).

lfos commented:
I had a closer look at the patches and they mostly look good to me. I am not sure about the patch dropping time ranges. We should allow for, say, listing all appointments starting from 3 p.m. to 10 a.m. on the next day.

The `general.systemdialogs` thing is weird and we should definitely debug that instead of simply turning it on. Maybe it is because something is now unexpectedly printed in CLI mode with system dialogs turned on.

What I meant in the previous comment is that we should also extend the test cases and add tests for the new "inclusive date range" behavior.

You can build an HTML version of the man page by using asciidoc directly on the text file.

lhca commented:
> I had a closer look at the patches and they mostly look good to me. I am not sure about the patch dropping time ranges. We should allow for, say, listing all appointments starting from 3 p.m. to 10 a.m. on the next day.

It's true that a time of day is no longer accepted as part of a date, but nothing was dropped. What you wish may be desirable, but it is not possible in 4.3.0.
 
> The `general.systemdialogs` thing is weird and we should definitely debug that instead of simply turning it on. Maybe it is because something is now unexpectedly printed in CLI mode with system dialogs turned on.

It's the other way around. The seven ical tests expect an import report like

> Import process report: 0017 lines read
1 app / 0 events / 1 todo / 0 skipped

This is only output when systemdialogs is on, and all ical tests failed when I ran the tests. The Travis build still succeeds, so apparently it succeeds, whether systemdialogs is on or off. I have no way of debugging that. The function `show_dialogs()` is a wrapper for `conf.system_dialogs` (=`general.systemdialogs`), and used to control whether save-messages were displayed. By intention the status bar messages about save and reload cannot be suppressed, so import reports are the only output controlled by `general.systemdialogs`
> 
> What I meant in the previous comment is that we should also extend the test cases and add tests for the new "inclusive date range" behavior.

Definitely.
 
> You can build an HTML version of the man page by using asciidoc directly on the text file.

That was not the problem I had in mind. The Makefile has a target for that. But I could find no way of _publishing_ the result for comments.

lhca commented:
> > I had a closer look at the patches and they mostly look good to me. I am not sure about the patch dropping time ranges. We should allow for, say, listing all appointments starting from 3 p.m. to 10 a.m. on the next day.
> 
> It's true that a time of day is no longer accepted as part of a date, but nothing was dropped. What you wish may be desirable, but it is not possible in 4.3.0.

After thinking more about this: you are right, and what you ask for _is_ possible in 4.3.0 (with the proper filter options). My confusion comes from the words "time ranges". I prefer "time of day" (Uhrzeit).

Let me try to clarify my own thoughts. Command line input dates are used in two different contexts:

- to specify a day range for output from queries (`--from`, `--to`, `--days`)
- to specify start and end times in filter options (`--filter-start-from`, etc.)

In the first context a time of day should be ignored and only the date (Datum) taken into account. In the second context the date with an optional time of day should be accepted.

Is this meaningful? Do you agree?

lfos commented:
IIRC, there is an important difference between the filter options and the query range options. The filters are applied when the items are loaded. They can only be used to filter based on item properties but not on occurrence properties. For example, `--filter-start-from` would not give you any future occurrences of a recurrent item if the first occurrence of that item is before the filter date.

The query range options can be used to filter occurrences. So it would be nice to have "time of day" support for them as well.

It's been quite a while since I have been working on this, so you might want to double-check.

lhca commented:
Reintroducing time of day in the input date format is of course easy enough. Supporting it in both contexts mentioned above is another matter.

The filter option context will require different handling of date only and date with time of day (because the first is a time span of 24 hours, the second a point in time). But that should be doable, and is then an enhancement of behaviour already present. I will work on that.

As for the query range context see below.

> IIRC, there is an important difference between the filter options and the query range options. The filters are applied when the items are loaded. They can only be used to filter based on item properties but not on occurrence properties. For example, `--filter-start-from` would not give you any future occurrences of a recurrent item if the first occurrence of that item is before the filter date.

Thank you. That's also my understanding. The new man-page tries to make the difference clear with a similar example (in the box "Start and end time of events and recurring items"). 

> The query range options can be used to filter occurrences. So it would be nice to have "time of day" support for them as well.

The man-page example similarly mentions that the query range may be used to limit the output. But requiring that it should work not just for dates, but also for dates with time of day (granularity of seconds) is new. This is definitely _not_ possible in 4.3.0, so there is no question of anything being dropped here.

Support for this is a major effort in my estimation. The input date is used directly to loop through the query range by increments of a day, and is passed on as argument to the `day_store_*()` functions. I can't assess whether it's possible or meaningful at all.

lfos commented:
Ok, I'd say updating the man page is fine then. We can improve this in a future version of calcurse. Will have a closer look at all the patches soon. Thanks for all the work!

lfos commented:
Did not review in full detail but merged everything into `pu`. Will have a closer look at the code and man page changes soon. Looks very promising, thanks for working on this!

lfos commented:
I decided to reopen this until it is merged into master. An HTML version of the (current version of the) new man page can be found under [1]. I will review that and comment here.

[1] https://calcurse.org/files/calcurse.1-pu.html

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+DESCRIPTION
+-----------
+
+Calcurse is a calendar and scheduling application for use in a terminal
```
I know that this has just been copied from the previous revision of the file but I wonder whether we should always use lower case for the proper name `calcurse`, even if it is at the beginning of a sentence. Opinions?

lhca replied:
Corrected.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
 
-The following options are supported:
+.Input and Output Date Format
+*****************************
```
I'd skip that heading. It may look nice and consistent in the generated HTML but the primary purpose of this document is to serve as a man page and the additional heading is very uncommon there.

lhca replied:
I admit that I mainly tried to make the HTML version look nice at the expense of the traditional terminal version. I have long since lost contact with the day-to-day IT world. Do (Unix) people still read man pages with the man command? Anyway, I have gone back to a traditional layout with just a few section levels (three).

lfos replied:
Yes, I believe that people still rely on man pages, especially when looking for a specific command line option. Of course, Google often helps as well but I'd say man pages are still being used, especially among heavy shell users (which is one of the main target groups of calcurse).

Beyond that, this is the source code for the man page. If we decide that we want to drop the man page and replace it with HTML documentation, we should do that instead of keeping the man page and making it ugly, while trying to make it convert into a pretty HTML page .

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+
+[[_limiting_options]]
+.Filter, format and day range options
+*************************************
```
Again, please make this a proper level-1 heading as it was before. While the new structure is semantically "correct", we should focus on making things easily visible in the man page. We should also avoid having many subheadings which make the man page harder to read.

lhca replied:
Corrected.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+  Invert the combined effect of any other filters, i.e. load the items that
+  do 'not' match them.
+
+.'general filters'
```
Those subheadings in particular look very strange in the new man page. The formatting of the `general filters` heading is different from the `calendar filters` heading.

lhca replied:
Corrected.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
-*S*::
-  Print the start time of the appointment using the *hh:mm* format
-*d*::
+[horizontal]
```
These commands generate labelled tables in the generated man page which look quite odd...

lhca replied:
Corrected.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
 
-Files
+[[_files]]
+FILES
```
From a cursory read, I really like the updates to the files section and everything below. I will have a closer look soon!

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+and issue notifications; it stops automatically when the interactive mode is
+reentered.
+
+This man-page mainly describes the command-line mode.
```
"man page", no hyphen.

lhca replied:
Corrected.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
-Description
------------
+--
+*calcurse* [-D 'directory'] | [-C 'confdir'] | [-c 'calendar-file']
```
`datadir` instead of `confdir`?

lhca replied:
?? I considered `datadir` instead of `directory`.

lfos replied:
Yes, that's what I meant, sorry.

lhca replied:
I retained _directory_, because it contains both data and configuration files when `-D` is used alone.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+.Input and Output Date Format
+*****************************
+Many options require a 'date' argument, and query results per day are set apart 
+by a leading 'date line'.
```
I see what you intended with this introductory section but I am not sure whether it is all that useful. A man page is used like a dictionary most of the time and rarely read from cover to cover. Maybe we should move all this to the formatting section? As a user, I'd expect to find it there.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+Filter options have effect on queries (*-Q* and query short-forms), grep
+(*-G*), purge (*-P*) and export (*-x*). Format options have effect on queries,
+grep and *--dump-imported*. Day range options have effect on queries only.
+*************************************
```
Again, I do not think this is very useful in practice. Let's rather turn all subsections into proper sections and move the pieces of information from this box to the intros of the corresponding new sections (where not present already).

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+scripting.
+
+*%d* ~ *%(duration*`[`*:*'format'`]`*)*::
+*%r* ~ *%(remaining*`[`*:*'format'`]`*)*::
```
I see you intentation here, but even as a calcurse developer I was confused by this syntax. Can we simply drop the short versions here and add something like "%(duration) is the long version of %d and %(remaining) is the long version of %r" instead?

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+  must precede *E*, 4) no other flags or width specifications are supported
+
+*%s* ~ *%(start*`[`*:*'format'`]`*)*::
+*%e* ~ *%(end*`[`*:*'format'`]`*)*::
```
Same here. I think we can simply drop `*%s* ~` and `*%e* ~` here since they are already mentioned in the text below.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
 
-The following structure is created in your $HOME directory (or in the directory
-you specified with the *-D* option), the first time calcurse is run:
+The following structure is created by default in your $HOME directory the
```
Maybe we can change "$HOME directory" to "home directory" while touching these lines...

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+----
+$ calcurse
+
+$ calcurse -C $HOME/.calcurse/config
```
This should be quoted: `calcurse -C "$HOME/.calcurse/config"`.

lfos replied:
If the home directory contains any space, the command without the quotes breaks. It's a corner case but it does not hurt to convert the example into a proper command that works under all circumstances.

lhca replied:
Corrected. Spaces crept into directory names a long time ago. Sigh.

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+wanted, they can, of course, be removed interactively. If there are many, it
+can be a tedious task and may be done better as in the following two examples.
+'Note.' The filter options specify what is retained, not what is removed!
+Presently it is not possible to list the items to be removed.
```
Is this true? Can't we use `-G --filter-invert`?

lfos commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
 ----
 
-calcurse - text-based organizer
+calcurse - terminal-based organizer for interactive and command line use
```
Should we come up with a good concise description to use everywhere? Right now, the man page, documentation, the web site and distribution packages each use their own descriptions.

It should make the core features obvious, include the most relevant search keywords (calendar, scheduling, command line) and be as short as possible...

lfos replied:
I thought of this before and the only thing I do not like about it is that it's too long. I think we should aim for something with <80 characters but I cannot think of anything better right now.

lhca replied:
<80 is a tall order. This one is 77 characters long:

calcurse - customizable calendar app with reminders and command line features

Here 'interactive use'  is understood as opposed to  'command line'. I believe that 'app' is in common use for 'program' or 'application'.

lhca commented on doc/calcurse.1.txt:
```diff
@@ -30,519 +30,714 @@
+*calcurse* -a | -d 'date' | -d 'number' | -n | -r['number'] | -s['date'] | -t['number']
+
+*calcurse* -h | --status | -G | -g | -i 'file' | -P | -x['file'] | --daemon
+--
```
Synopsis section.

The first line is obviously wrong. Corrected.

I don't agree as to the other lines. The query option (`-Q`) stands apart. It is the command line equivalent of the interactive interface and is the only option for which the day range concept is applicable and definitely not on a par with `-P` and `-G`. The third line lists all the various query short forms.

The last line lists the remaining options.

It is an attempt to give a quick overview of a mixed bag of options.

lfos replied:
I agree that `-Q` is special. However, these lines also help increase the visibility of the different modes of operation. I don't feel strongly about it but I'd personally prefer having `-G` and `-P` grouped in a separate line for that reason. If that does not convince you, we can keep it like that.

lhca replied:
They have got their own line.

lhca replied:
I think that your suggestion is an improvement, but it's not in pu. Didn't you like it?

lhca replied:
> I think that your suggestion is an improvement, but it's not in pu. Didn't you like it?

My mistake. I just realized that the last round of changes didn't make it into my local CLI-update branch before the squashing and pushing (grr and some danish swearing).

lfos commented:
Could you also try to rebase this branch on master (or pu) while reworking it? There only is a small conflict in `args.c`. That would make it much easier for me to merge and test stuff. I will give you some time to discuss and incorporate the suggested changes and later build and upload a new HTML version of the suggested man page.

Thanks!

lhca commented:
> Could you also try to rebase this branch on master (or pu) while reworking it? There only is a small conflict in `args.c`.

Ah, yes. I didn't notice that. It's the `io_init()` rewrite overtaking the CLI-update into pu. I have pushed an update.

lfos commented:
The io_init() patch has been merged to master already, so we cannot amend that. This branch needs to be rebased on master instead.

lfos commented:
While interactively rebasing, you could also squash the man page review updates into the original man page patch :)

lhca commented:
> While interactively rebasing, you could also squash the man page review updates into the original man page patch :)

I'm still not quite at ease with rebasing. Here is what I have done. I removed the `io_init()` stuff from the CLI-update branch and squashed all the man page commits into the original patch. Then I forced-pushed everything to GitHub without rebasing. Will that do?

lfos commented:
Thanks! I merged everything to pu with minor modifications, rebuilt and uploaded the HTML version. I like the outcome. Will merge to master soon unless there are any objections.

lfos commented:
I added most (but not all) of your updates to pu. Thanks!

lfos commented:
Merged into master.
